File: code/backend/php/src/Core/Port/Story/StoryCrudPort.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Port\Story;

/**
 * StoryCrudPort — Step 17: Inbound port for generic admin CRUD on story entities.
 */
interface StoryCrudPort
{
    /**
     * @return array[]|null  List of entities, or null if story not found.
     */
    public function listEntities(string $storyUuid, string $entityType): ?array;

    /**
     * @return array|null  Entity data, or null if not found.
     */
    public function getEntity(string $storyUuid, string $entityType, string $entityUuid): ?array;

    /**
     * @return array|null  Created entity, or null if invalid.
     */
    public function createEntity(string $storyUuid, string $entityType, array $data): ?array;

    /**
     * @return array|null  Updated entity, or null if not found.
     */
    public function updateEntity(string $storyUuid, string $entityType, string $entityUuid, array $data): ?array;

    /**
     * @return bool  True if deleted, false if not found.
     */
    public function deleteEntity(string $storyUuid, string $entityType, string $entityUuid): bool;

    /**
     * @return array|null  Created story.
     */
    public function createStory(array $data): ?array;

    /**
     * @return array|null  Updated story, or null if not found.
     */
    public function updateStory(string $storyUuid, array $data): ?array;
}

File: code/backend/python/app/core/ports/story/story_crud_port.py
"""
StoryCrudPort — Inbound port for Step 17 generic admin CRUD.
Provides create, read, update, delete for all story sub-entities.
"""
from abc import ABC, abstractmethod
from typing import Dict, Any, List, Optional


class StoryCrudPort(ABC):
    @abstractmethod
    def list_entities(self, story_uuid: str, entity_type: str) -> Optional[List[Dict[str, Any]]]:
        """List all entities of a type within a story. Returns None if story not found."""
        pass

    @abstractmethod
    def get_entity(self, story_uuid: str, entity_type: str, entity_uuid: str) -> Optional[Dict[str, Any]]:
        """Get a single entity by UUID. Returns None if not found."""
        pass

    @abstractmethod
    def create_entity(self, story_uuid: str, entity_type: str, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create a new entity. Returns None if story not found or data invalid."""
        pass

    @abstractmethod
    def update_entity(self, story_uuid: str, entity_type: str, entity_uuid: str, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update an entity. Returns None if not found."""
        pass

    @abstractmethod
    def delete_entity(self, story_uuid: str, entity_type: str, entity_uuid: str) -> bool:
        """Delete an entity. Returns True if deleted, False if not found."""
        pass

    @abstractmethod
    def get_story(self, story_uuid: str) -> Optional[Dict[str, Any]]:
        """Get a single story by UUID. Returns None if not found."""
        pass

    @abstractmethod
    def create_story(self, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create a new story."""
        pass

    @abstractmethod
    def update_story(self, story_uuid: str, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update story metadata. Returns None if not found."""
        pass
